### ISSUE TITLE
Inconsistent type deduction between `final x = …;` and `for (final x in …)`.

### ISSUE THREAD

**Issue by Cyp** [dart-lang/language#1866]
Not sure if this is an SDK issue or a language issue, but `t1` is semantically identical to `t2`, and `t3` is identical to `t4`, so it doesn't make sense that only one of each works. Neither with nor without a temporary variable seems to work in all cases.

```dart
List<int>? arr;

// Expected: x is int. OK.
int? t1() {
  final xs = arr ?? const [];
  for (final x in xs) {
    return x;
  }
}

// Expected: x is int. Actual: x is dynamic!
int? t2() {
  for (final x in arr ?? const []) {
    // A value of type 'dynamic' can't be returned from the function 't2' because it has a return type of 'int?'.
    return x;
  }
}

// Expected: x is num. Actual: x is int!
num? t3() {
  // The element type 'double' can't be assigned to the list type 'int'.
  final xs = arr ?? const [3.14];
  for (final x in xs) {
    return x;
  }
}

// Expected: x is num. OK.
num? t4() {
  for (final x in arr ?? const [3.14]) {
    return x;
  }
}
```

Dart SDK version: 2.13.4 (stable)

**Comment by Cyp** [comment]
I'd lean towards `const []` meaning `const <Never>[]` and `[]` meaning `<Never>[]` (since `Never` is the common supertype of all the elements in the list).

And with a special case for backwards compatibility that if a declared variable would have end up with type `SomeType<Never>` and the `Never` wasn't explicitly specified, that the derived `Never`s in the variable type recursively turn into `Object?`/`dynamic`s.

**Comment by lrhn** [comment]
> If I understand @lrhn's idea correctly, I believe he's suggesting that the context for the RHS should always be the static type of the non-null version of the static type of the LHS, regardless of the outer context type.

I wouldn't actually do that. If we use the type of the LHS, it should be made non-nullable. I don't necessarily think we should use it at all.

I'm guessing the problem we're trying to solve by using the static type of the LHS is exactly the `intListOrNull ?? []` code.
If we always use the context type for `[]`, and ignore the type of `intListOrNull`, we will *too often* end up with `List<Object?>`/`List<dynamic>`, basically every time there is no context type.

However, if we use the LHS type as a *requirement* then `intListOrNull ?? [0.5]` won't infer `List<num>`, it just rejects `[0.5]` as not a `List<int>`.

So, if anything, we need to use the LHS type as a *hint*, not a *requirement*.

What I suggest is:
* If there is a context type *C* for `LHS ?? RHS`, use *C*? as context type for `LHS`, getting static type *S*, and *C* as context type for `RHS` getting static type *T*. 
* If there is no context type for `LHS ?? RHS`, let *S* be the static type of `LHS` inferred with no context type. Let *T* be the static type of `RHS` inferred with context type **NON_NULL**(*S*). (It's not an error if *T* is not a subtype of **NON_NULL**(*S*), the context type is just a hint).
* In either case, the static type of `LHS ?? RHS` is **UP**(*S*, *T*).

It's not an error for this to not be a subtype of the context type *C* unless the context requires a value of that type (like the second operand of `??` *doesn't*).


---
(I also think we should use the cast type as context type for a cast expression, so `e as T` gives `e` a context type of `T`, as a *hint* on the sub-expression.)

**Comment by lrhn** [comment]
> (Nit: I think you need that last line to be **UP**(**NON_NULL**(*S*), *T*).

Good catch, I originally had it but managed to throw that away at some point. Fixed.

About the goal: True, my idea wouldn't fix the problem. I'm not absolutely sure I definitely want to fix the problem, not if the rules become too special-casey, but if we can, then ... it'll probably help users.

The problem here is that I can't find a way to do that which won't prevent `intListOrNull ?? [0.5]`. That's perfectly good code, and should have type `List<num>`.

But, if we use `List<int>` as context type for `[0.5]`, then it becomes `<int>[0.5]` and then fails, because we always favor the context type during downwards inference..

If we don't use `List<int>`, but just `Iterable<?>`, then `intListOrNull ?? []` will likely become `... ?? <dynamic>[]`.

It seems we have to choose one annoyance or the other. 

The only solution that can avoid both annoyances, that I can see, is to somehow pass `List<int>` in as a context type *hint*, where we then use `int` as context type for the elements (if any) of the list literal, but we don't choose to make the literal a `List<int>` until we have actually *seen* the elements. If they all agree, then by all means make it a `List<int>`. If not, make it a list that can contain its actual elements, then worry about whether that fits into the context later.

(Could we actually do that for *all* downwards inference? Not lock-in the type arguments immediately on the way down, but just use the context type as hints, propagate the hints into context types for sub-expressions, then build the actual type on the way up? If we *can* use the hint type's type arguments with the actual sub-expression types, we do so, otherwise we infer type arguments from the sub-expression types as if there was no context type.)

**Comment by leafpetersen** [comment]
I think the general reconstruction here is correct.  I vaguely remember the Dart 2 discussions about this, and I think we chose to use the context type if we had it, but otherwise to heuristically choose the LHS type as the context for the RHS precisely because it made a lot of existing code compile.  As for a broader rewrite of the inference strategy... we can push this on the stack of things to consider, but I think it needs to some careful design.
